---
title: "Reproducible script for the publication"
author: "Weinholdt Claus"
date: "2018-24-4"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
Weinholdt et. al __Analysis of genes regulated by isoforms of the epidermal growth factor receptor in a glioblastoma cell line__

## Contents

* [Getting started](#start)
    * [Installation](#Installation)
    * [Data](#data)
* [Analysis](#Analysis)
* [Comapre Illumina data with RT-qPCR](#Comapre)
    * [Examples for group c](#GrC)
    * [Fold changes plot of Illumina data vs RT-qPCR](#FCplot)

## <a name="start"></a>Getting started

### <a name="Installation"></a>Installation of necessary packages

Install and load packages containing the functions for the analyses

```r
install.packages("devtools")
devtools::install_github("BGSC")
library(BGSC)
```

### <a name="data"></a> Load expression data
<!-- Loading data  -->
```r
data(ExpData)
```

```{r, comment=NA, echo=FALSE,}
suppressWarnings( library("BGSC",quietly = TRUE,warn.conflicts = FALSE) )
library(ggplot2)
library(gridExtra)
library(grid)
library(gtable)
suppressWarnings( library(purrr,quietly = TRUE,warn.conflicts = FALSE) ) 
```

## <a name="Analysis"></a> Run analysis

### Normalizing Illumina BeadChips expression data

We use the function *neqc* function from the *limma* package which is developed for normalizing Illumina BeadChips.
*neqc* performs background correction using negative control probes followed by quantile normalization using negative and positive control probes. 
The *Illumina GenomeStudio* calculates and reports a detection p-value, which represents the confidence that a given transcript is expressed above the background defined by negative control probes.
For further analysis, we used only the those probes for which the detection p-values for all six probes were below 0.05.
```{r normalizeExpData}
normData <- normalizeExpData()
```

### Calulating the log likelihood for each gene in each class (*a, b, c, and d*)
We define that:
* Genes of group $a$ are never regulated by EGF, whereas genes of groups $b-d$ are regulated by EGF.
* Genes of group $b$ are regulated by EGF only through other receptors besides EGFR isoforms.
* Genes of group $c$ are regulated by EGFR isoforms II - IV and not by other receptors.
* Genes of group $d$ are regulated by EGFR isoform I and not other receptors or EGFR isoforms II - IV.
<!--  Based on this reduction, we can now formulate the goal of this work as the identification of putative target genes regulated by EGFR isoforms II - IV and not by other receptors or more crisply as the goal of identifying genes of group $c$. -->


```{r logLik}
Lsets <- get.Lset()
normDataLogLik <- logLikelihoodOfnormData(normData$E)
```

#### Density plots with fitted normal distribution 
```{r Density, echo=TRUE, fig.height=9, fig.width=9}
GeneExample <- c('ILMN_1687840','ILMN_1684585','ILMN_1730999','ILMN_2320964') 
names(GeneExample) <- c('a','b','c','d')
tmpPlot <- purrr::map2(GeneExample,names(GeneExample),function(.x,.y) Density.NV.fit.plot(id = .x ,normData,useGroup = .y ,DOplot = FALSE) )
grid.arrange(tmpPlot$a + theme(legend.position = "none"),
             tmpPlot$b + theme(legend.position = "none"),
             tmpPlot$c + theme(legend.position = "none"),
             tmpPlot$d + theme(legend.position = "none") ,ncol=2,nrow=2)
      
```

### calulate BIC from logLik 
```{r BIC}
  npar <- sapply(Lsets, function(x) sum(!sapply(x,is.null ) )) + 1  ## number parapeters for LogLilk -> mean + var 
  k <-  sapply(Lsets, function(x) sum( sapply(x,length) ))
  normDataBIC <- get.IC(normDataLogLik , npar, k , IC = 'BIC')
  BICminInd <- apply( normDataBIC,MARGIN = 1, FUN = minIndex)
  print(table(BICminInd))
``` 

###  calulate Posterior from BIC
```{r Posterior}
  normDataPosterior <- get.Posterior( normDataBIC ,Pis = c(0.7,0.1,0.1,0.1))
  POSTmaxInd <- apply(normDataPosterior, MARGIN = 1 ,FUN = maxIndex)
  print(table(POSTmaxInd))
```

###  get gene classes
```{r PostClass}   
  PostClass <- get.gene.classes(data = normDataPosterior,indexing = "max",filter = 0.75, DoPlot = TRUE)
```   

## <a name="Comapre"></a> Comapre Illumina data with RT-qPCR

### <a name="GrC"></a> Examples for group c

```{r load qPCR}
    MeanFoldChangeClass <- get.log2Mean.and.log2FC(normData = normData)
    qCPRdataC <- getQPCR()
    
    IDs.dt <- data.table::data.table(normData$genes,keep.rownames = T,key = 'rn')
    IDs.dt.c <- IDs.dt[rownames(PostClass$resFilter$c),]
    data.table::setkey(IDs.dt.c,'SYMBOL')
    
    qgenesIDs <- lapply(qCPRdataC$rn, function(qg) as.character(IDs.dt.c[qg,][['rn']]) )
    names(qgenesIDs) <- qCPRdataC$rn
  
    print(do.call(c,qgenesIDs))
```

### Schematic expression patterns of group c 

c0 is red and c1 is blue ... 

```{r group c example, echo=FALSE, fig.height=4, fig.width=5}
      ### bar log2 FC qPCR Illumina  ----------------------------------------------------------------------
        Leset <- get.Lset()
        pmat <- t(matrix(c(0,1,0,0,0,1),2,3))
        dimnames(pmat) <- list( c('no RNAi','RNAi by siRNA ALL','RNAi by siRNA I'),c('no EGF','EGF'))  
        pmat <- pmat[c(1,3,2),] ## to get same oder as in paper
        bk = seq(0,max(1),by = 1)
        #hmcols <- c(RColorBrewer::brewer.pal(9,"Reds")[6],RColorBrewer::brewer.pal(9,"Blues")[7])# colorRampPalette(c("blue", "red"))(length(bk))
        hmcols <- RColorBrewer::brewer.pal(8,'Paired')[c(6,2)] 
        pheatmap::pheatmap(pmat,color = hmcols,breaks = seq(-1,1,by = 1) ,display_numbers = T,fontsize_number = 20,number_format = '%.0f',cluster_rows = FALSE,cluster_cols = FALSE,main='expression pattern group c',fontsize = 10,gaps_row = c(1,2),gaps_col = 1,border_color = 'black',number_color = 'black',legend = FALSE)
```

###  Expression patterns 
```{r group c exp genes ,echo = FALSE }                  

        Gene.gtable <- list()
        for(tmoG in names(qgenesIDs)){
          pmat <- t(matrix(normData$E[qgenesIDs[[tmoG]],],2,3))
          dimnames(pmat) <- list( c('no RNAi','RNAi by siRNA ALL','RNAi by siRNA I'),c('no EGF','EGF'))  
          pmat <- pmat[c(1,3,2),] ## to get same oder as in paper
          
          pmat <- 2^pmat
          tmpnScore <-  ceiling(pmat/100)*100
          # tmpnScore <-  ceiling(pmat/10)*10
          
          bk = seq(0,max(tmpnScore),by = 1)
          hmcols <- colorRampPalette(c("white", "darkred"))(length(bk)-1)
          #pheatmap::pheatmap(pmat,color = hmcols,breaks = seq(0,max(tmpnScore),by = 1) ,display_numbers = T,fontsize_number = 12,cluster_rows = FALSE,cluster_cols = FALSE,main=tmoG,fontsize = 12,gaps_row = c(1,2),gaps_col = 1,border_color = 'black',number_color = 'black', )
                             # filename = paste0(plotDir,'/Heatmap_Score_',tair,'.pdf') ,width = 10 ,height = 10)
          Gene.gtable[[tmoG]] <- pheatmap::pheatmap(pmat,silent = T,color = hmcols,breaks = seq(0,max(tmpnScore),by = 1) ,display_numbers = T,fontsize_number = 15,cluster_rows = FALSE,cluster_cols = FALSE,main=tmoG,fontsize = 12,gaps_row = c(1,2),gaps_col = 1,border_color = 'black',number_color = 'black')
        }
```

```{r group c exp genes plot, echo=FALSE, fig.height=8, fig.width=12}
  addBorder_gtable <- function(g){ g <- gtable_add_grob(g,
                                            grobs = rectGrob(gp = gpar(fill = NA, lwd = 1)),
                                            t= 1 , b = nrow(g), l = 1, r = ncol(g))
                      # g <- gtable_add_grob(g,
                      #                      grobs = rectGrob(gp = gpar(fill = NA, lwd = 1)),
                      #                      t = 1, l = 1, r = ncol(g))
                      
                      return(g)
        }
  grid.arrange(
    addBorder_gtable(Gene.gtable$CKAP2L$gtable),
    addBorder_gtable(Gene.gtable$ROCK1$gtable),
    addBorder_gtable(Gene.gtable$TPR$gtable),
    addBorder_gtable(Gene.gtable$ALDH4A1$gtable),
    addBorder_gtable(Gene.gtable$CLCA2$gtable),
    addBorder_gtable(Gene.gtable$GALNS$gtable),
    nrow=2,ncol=3)             
```

### Barplot mean expression data of Illumina 
By summarizing the the expression of c0 and c1 we get ... with std-err

```{r barplot exp data ,echo=FALSE}       
    ### bar Illumina expression ----------------------------------------------------------------------
      PlotDataEXP <- make.plot.data.exp.Ill.qPCR(qCPRdata = qCPRdataC, MeanFoldChangeClass = MeanFoldChangeClass,class="c")
      
      # print(PlotDataEXP)
      
      rns <- levels(PlotDataEXP$Gene)[c(2,5,6,1,3,4)] ; PlotDataEXP$Gene <- factor(PlotDataEXP$Gene, levels = rns)
      rns <- levels(PlotDataEXP$pid)[c(2,5,6,1,3,4)]  ; PlotDataEXP$pid <- factor(PlotDataEXP$pid, levels = rns)
      
      cols <-  RColorBrewer::brewer.pal(8,'Paired')[c(6,2)] 
      limits <- aes(ymax = PlotDataEXP$Mean + PlotDataEXP$stderr , ymin=PlotDataEXP$Mean - PlotDataEXP$stderr)
      dodge <- position_dodge(width = 0.9)
      
      base_size <- 12
      PlotExp <- ggplot(PlotDataEXP,aes(x=factor(Gene),y=Mean,fill=factor(Set))) +  
        geom_bar(position="dodge",stat="identity") + 
        geom_errorbar(limits, position=dodge, width=0.25) +
        ylim(c(0 ,10)) +
        scale_fill_manual(values = cols,name="") + 
        labs(x = "", y = "Log2 mean expression",title="mean expression data of Illumina") +  
        thememapBarplot(base_size = base_size,legend_key_size = 0.6) + 
        # theme( aspect.ratio = 9 / 16) + 
        theme(legend.position="bottom") 
      print(PlotExp) 

```

<!-- 
```{r results="PlotDataEXP", echo = FALSE}
knitr::kable(PlotDataEXP, caption = "Table 1", row.names = F)
```
-->

### <a name="FCplot"></a> Log2 fold changes of Illumina data vs RT-qPCR
```{r barplot FC, echo=FALSE}
    ### bar log2 FC qPCR Illumina  ----------------------------------------------------------------------
      PlotDataFC <- make.plot.data.FC.Ill.qPCR(qCPRdata = qCPRdataC, MeanFoldChangeClass, class="c" )
      
      # print(PlotDataFC)
      
      rns <- levels(PlotDataFC$Gene)[c(2,5,6,1,3,4)] ; PlotDataFC$Gene <- factor(PlotDataFC$Gene, levels = rns)
      rns <- levels(PlotDataFC$pid)[c(2,5,6,1,3,4)]  ; PlotDataFC$pid <- factor(PlotDataFC$pid, levels = rns)
      
      cols <- RColorBrewer::brewer.pal(11,"PRGn")[c(2,10)]
      limits <- aes(ymax = PlotDataFC$FC + PlotDataFC$stderr , ymin=PlotDataFC$FC - PlotDataFC$stderr)
      dodge <- position_dodge(width=0.9)
      
      # g1 <- ggplot(PlotDataFC,aes(x=factor(pid),y=FC,fill=factor(Set))) +
      PlotFC <- ggplot(PlotDataFC,aes(x=factor(Gene),y=FC,fill=factor(Set))) +  
        geom_bar(position="dodge",stat="identity") + 
        geom_errorbar(limits, position=dodge, width=0.25) +
        ylim(c(-2 ,2)) +
        labs(x = "", y = "Log2-fold change",title="Comparison of Illumina data and RT-qPCR") +
        scale_fill_manual(values = cols,name="") + 
        thememapBarplot(base_size = base_size,legend_key_size = 0.6) + 
        # theme( aspect.ratio = 9 / 16) + 
        theme(legend.position="bottom") 
      print(PlotFC)     
```

```{r results="PlotDataFC", echo = FALSE}
knitr::kable(PlotDataFC, caption = "Table 2", row.names = F)
```







Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
